subtests/docker_cli/info/info.py

```diff
@@ -40,38 +41,81 @@ def run_once(self):
-        vals = []
-        for x in out:
-            try:
-                vals.append(x[1].strip())
-            except IndexError:
-                vals.append(None)
-        return dict(zip(keys, vals))
+        """
+        'docker info' returns a human-readable list of key: value pairs.
+        Some are indented by a space, indicating that these key/values
+        are subelements of a previous element. To wit:
+
+             Images: 3
+             Server Version: 1.12.6
+             Storage Driver: devicemapper
+              Pool Name: vg--docker-docker--pool
+              Pool Blocksize: 524.3 kB
+              ...
+             Logging Driver: journald
+
+        In this case 'Images', 'Server Version', and 'Logging Driver'
+        are simple tuples but 'Storage Driver' has both a value and
+        a set of further key/value pairs: Pool Name, Pool Blocksize, ...
+
+        We parse that and return a dict with the expected key/value
+        mapping *and* an extra: for all elements with subelements,
+        'element...' (element name plus three dots) is a dict containing
+        the subelements. E.g. x['Storage Driver...']['Pool Name'] = 'vg--etc'
+        """
+        out = {}
+        current_key = None
+        for line in cli_output.split('\n'):
+            # Almost every line will be Foo: Bar, but 'Insecure Registries:'
+            # is followed by a simple list of IPv4 netmasks
+            if ': ' in line or line.endswith(':'):
+                (key, value) = [e.strip() for e in line.split(':', 1)]
+            else:
+                (key, value) = (line.strip(), '')
+            if line.startswith(' '):
+                if not current_key:
+                    raise IndexError("sdfsdf")
+                if current_key not in out:
+                    out[current_key] = {}
+                out[current_key][key] = value
+            else:
+                out[key] = value
+                current_key = key + '...'
+        return out
 
     def postprocess(self):
         # Raise exception on Go Panic or usage help message
         outputgood = OutputGood(self.stuff['cmdresult'])
         info_map = self._build_table(outputgood.stdout_strip)
-        # Verify some individual items
-        self.failif_ne(info_map['Storage Driver'].lower(), 'devicemapper',
-                       'Storage Driver')
-        self.failif_ne(info_map['Data file'].lower(), '',
-                       'Data file')
-        self.failif_ne(info_map['Metadata file'].lower(), '',
-                       'Metadata file')
+
+        # We support multiple storage drivers. Each one has a different
+        # set of key/value settings under 'info'; so each one has a
+        # dedicated helper method for validating.
+        storage_driver = info_map['Storage Driver'].lower()
+        try:
+            handler = '_postprocess_' + storage_driver
+            getattr(self, handler)(info_map['Storage Driver...'])
```

Comment:
My goal was, and always is, to avoid unnecessary unnecessary duplication. A secondary goal is to avoid long `if-else`s. With my scheme, to add support for a new back end, you add a new function (nothing else). With the above one, you need to (1) add a new function, (2) add a new `elif` and make sure you get the string right, (3) add a call to the new function. And you get a hard-to-read `if` block as well. I'm a big fan of table-driven programming where there's fewer things to go wrong.

I don't understand your class-attribute-handler-etc comment but will spend more time looking into it. Thanks for the suggestion.